### PR TITLE
feat(dramatiq): add `dramatiq` integration

### PR DESCRIPTION
**Pull request by erikayasuda**
## Overview
This PR adds the [`dramatiq`](https://github.com/Bogdanp/dramatiq/tree/master) library as a supported integration for `dd-trace-py`. This addresses https://github.com/DataDog/dd-trace-py/issues/5043.

### In Scope for this PR
- Instrumenting the `dramatiq.actor.Actor.send_with_options()` method. This is the method called when a function with the `@dramatiq.actor` decorator is called asynchronously. See example below:

```python
# app.py
import dramatiq
from flask import Flask

app = Flask(__name__)

@dramatiq.actor
def my_func():
   return "response"

@dramatiq.actor
def my_other_func(a: int, b: int) -> int:
   return a + b

@app.route('/')
def index():
   my_func.send() # this calls send_with_options() under the hood
   my_other_func.send_with_options(args=(1, 1), options={"max_retries": 3})
   return 'hello world'

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=6060)
```

Running the above example Flask app with `ddtrace-run flask run` and hitting the `/` endpoint would generate a trace in the UI like the following:

Screenshot 2024-03-26 at 3 28 21 PM

The detailed span info for `my_other_func.send_with_options(...)` would look something like this:

```python
{
   actor: {
      name: my_other_func
      options: {"max_retries": 3}
   }
   env: test
   language: python
   span: {
      kind: producer
   }
}
```

#### NOTE
- The duration of the span is of the `send_with_options()` call itself, and not reflective of the execution duration of the function being asynchronously completed.
- When calling a function asynchronously with `send()`, it will also display on the UI as a `send_with_options()` span. This is because `send()` calls `send_with_options()` with empty options, so tracing both of these functions would create two spans for every instance that `send()` is called, with the exact same span information. 

### Out of Scope for this PR
- All other `dramatiq` methods.
- Supporting the actual duration of the function executed asynchronously by `send_with_options()`

The above and additional features can be investigated and added to this initial iteration at a later time.

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.
- [x] If change touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: tests/contrib/dramatiq/test_snapshots.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        patch()
+
+        @dramatiq.actor
+        def fn_task():
+            return "hi"
+
+        fn_task.send()
+        fn_task.send_with_options(options={"max_retries": 1})
+
+    @snapshot(wait_for_num_traces=0)
```
Comment (erikayasuda) [inline tests/contrib/dramatiq/test_snapshots.py:29]: This and line 42 for `test_fn_task_synchronous()`. This isn't actually doing what I want it to do, which is to confirm that there are 0 traces generated. It looks like `wait_for_num_traces` could accidentally prematurely consider this correct by meeting the loop exit condition, when in reality the call did generate a trace. Leaving this blank causes an indexing error though, so maybe this IS ok? Doesn't feel right though, so looking for suggestions on making this better.

File: tests/contrib/dramatiq/test_snapshots.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        @dramatiq.actor
+        def fn_task():
+            raise ValueError("test error")
+
+        fn_task.send()
+
+    # Ignoring these two values due to variance in method name
+    # Python 3.7 - 3.9 -> send_with_options
+    # Python 3.10+ -> Actor.send_with_options
+    @snapshot(ignores=["meta.error.message", "meta.error.stack"], wait_for_num_traces=1)
```
Comment (erikayasuda) [inline tests/contrib/dramatiq/test_snapshots.py:85]: This feels a little clunky, but not sure if there's a cleaner way to do this?

File: tests/contrib/dramatiq/test_snapshots.py (hunk 1 above)
Comment (emmettbutler) [reply tests/contrib/dramatiq/test_snapshots.py:30]: If you're wanting to check that no traces are generated, a snapshot test might not be the best fit. Instead, you could set up a `DummyTracer` and then verify that its list of traces was empty, similarly to [this example](https://github.com/DataDog/dd-trace-py/blob/fedf88e7161e7eff19038231c2b2fa8f8f75ddec/tests/contrib/cassandra/test.py#L441-L449).

File: tests/contrib/dramatiq/test_snapshots.py (hunk 2 above)
Comment (emmettbutler) [reply tests/contrib/dramatiq/test_snapshots.py:85]: The fact that `error.type` isn't ignored means this test still meaningfully verifies that an error was reflected in the span. I think it's fine to ignore the more specific details of that error.

File: tests/contrib/dramatiq/test_snapshots.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        # it should catch exceptions in task functions
+        @dramatiq.actor
+        def fn_task():
+            raise ValueError("test error")
+
+        fn_task.send()
+
+
+@snapshot(wait_for_num_traces=1)
+def test_fn_retry_exception(stub_broker, stub_worker):
```
[the remark this answers is not in the document]
Comment (emmettbutler) [reply tests/contrib/dramatiq/test_snapshots.py:98]: I've used the exact same approach. I can't find evidence in documentation, but fixtures to instance methods might be something that Pytest doesn't support.

**Review by emmettbutler — APPROVED** [review]

File: tests/contrib/dramatiq/test_snapshots.py (hunk 1 above)
Comment (erikayasuda) [reply tests/contrib/dramatiq/test_snapshots.py:30]: `Pin`!! 📌

**Pull request merged**